### PR TITLE
Add support for virtual cargo

### PR DESCRIPTION
**Pull request by SilentSpike**
Allows cargo objects to be added via classname and reduces simulation/synchronization overhead. Objects become physical upon unloading and cannot be re-made virtual, only scripts can add such virtual objects.

Also separates the player feedback from the loading/unloading code and adds a progress bar.

### CODE REVIEW

File: addons/cargo/XEH_postInit.sqf [hunk 1]
```diff
@@ -1,5 +1,45 @@
...
+    _itemName = getText (configFile >> "CfgVehicles" >> _itemClass >> "displayName");
+    _vehicleName = getText (configFile >> "CfgVehicles" >> typeOf _vehicle >> "displayName");
+
+    ["displayTextStructured", [[_hint, _itemName, _vehicleName], 3.0]] call EFUNC(common,localEvent);
+
+    if (_unloaded) then {
+        // Invoke listenable event
+        ["cargoUnloaded", [_item, _vehicle]] call EFUNC(common,globalEvent);
+    };
+
```
Comment (jonpas) [inline addons/cargo/XEH_postInit.sqf]: Wouldn't this all be nicer in separate functions rather than all in this file?

**Comment by SilentSpike** [comment]
These changes can likely be optimized/made neater. I'm not too familiar with the original code for cargo so it was a fairly messy workflow in order to get a functional implementation. Tested and working:

- Loading/unloading virtual objects
- Loading/unloading actual objects
- Cargo capacity with virtual/actual objects
- Progress bar and failure to load

**Comment by SilentSpike** [comment]
Wasn't entirely sure whether to separate the virtual/object handling between functions. It seemed like it would be wasteful to do that though.

Anyone more familiar with the code feel free to add your own commits to improve this implementation.

**Comment by SilentSpike** [comment]
In seperating the player feedback from the loading system:

- `fnc_startLoad` and `fnc_startUnload` initiate the progress bar (or fail immediately if the action can't be taken) which is based on the size of the item being moved
- The progress bar will raise the respective `"LoadCargo"` or `"UnloadCargo"` event (they were unused so I assumed this was to be their intended purpose)
- The event will take care of calling `fnc_(un)loadItem` and handle the player feedback based on whether it was successfully (un)loaded or not

So the loading and unloading functions themselves (the ones called by the events) contain no player feedback and are safe to use for scripting cargo manipulation

**Comment by jonpas** [comment]
> The progress bar will raise the respective "LoadCargo" or "UnloadCargo" event (they were unused so I assumed this was to be their intended purpose)

I actually added them as public API, we needed a way to trace it easily for our persistence system. But any use is of course great! :grin:

**Comment by SilentSpike** [comment]
Reassigning to commy since he was talking about messing with the cargo code similarly to this

**Comment by PabstMirror** [comment]
I'm thinking that addSpareParts should add virtual? (just note, doesn't have to be in this PR).

Will have some problems with #2593, so lets merge this first and I can work out the conflicts.

**Comment by jonpas** [comment]
addSpareParts uses events, so it's automatically virtual if I understand this PR correctly. I added AddCargoByClass event just because of addSpareParts.

**Comment by SilentSpike** [comment]
Yeah, it would automatically be made virtual by this PR. It could be changed to just call the actual function, but I didn't want to mess with that because I didn't know if there was any intentional structure to it

**Comment by jonpas** [comment]
> It could be changed to just call the actual function, but I didn't want to mess with that because I didn't know if there was any intentional structure to it

The intention was simply: it's cleaner if there are events, as events are made to be backward compatible (no name changes to it), while the functions below it can be changed as wished.

**Comment by Glowbal** [comment]
What is the status on this @SilentSpike ?

**Comment by SilentSpike** [comment]
It's done and waiting review :smile: (or if anyone wants to build on top of it, but so far nobody has spoken up).

**Comment by jonpas** [comment]
Except my first comment above, it looks good to me.

**Comment by commy2** [comment]
I'll merge this and other stuff after testing. Currently busy with BWMod though.

**Pull request merged**